pilot/pkg/serviceregistry/kube/controller/endpointslice.go

```diff
@@ -102,25 +104,22 @@ func (esc *endpointSliceController) onEventInternal(_, ep *v1.EndpointSlice, eve
 	} else {
 		esc.updateEndpointSlice(ep)
 	}
-	hostnames := esc.c.hostNamesForNamespacedName(namespacedName)
-	// Trigger EDS push for all hostnames.
-	esc.pushEDS(hostnames, namespacedName.Namespace)
 
 	name := serviceNameForEndpointSlice(esLabels)
 	namespace := ep.GetNamespace()
 	svc := esc.c.services.Get(name, namespace)
-	if svc == nil || svc.Spec.ClusterIP != corev1.ClusterIPNone || svc.Spec.Type == corev1.ServiceTypeExternalName {
+	if !serviceNeedsPush(svc) {
 		return
 	}
+	log.Infof("triggering EDS push for %s %s in namespace %s", name, event, namespace)
+
+	hostnames := esc.c.hostNamesForNamespacedName(namespacedName)
+	// Trigger EDS push for all hostnames.
+	esc.pushEDS(hostnames, namespacedName.Namespace)
 
 	configs := []types.NamespacedName{}
 	pureHTTP := true
 	for _, modelSvc := range esc.c.servicesForNamespacedName(config.NamespacedName(svc)) {
-		// skip push if it is not exported
-		if modelSvc.Attributes.ExportTo.Contains(visibility.None) {
```

Comment:
isn't this skipped in endpoint handler, is this a refactor

Reply:
I moved it above